[SPARK-13696] Remove BlockStore class & simplify interfaces of mem. & disk stores

JoshRosen opened the pull request:
Today, both the MemoryStore and DiskStore implement a common `BlockStore` API, but I feel that this API is inappropriate because it abstracts away important distinctions between the behavior of these two stores.

For instance, the disk store doesn't have a notion of storing deserialized objects, so it's confusing for it to expose object-based APIs like putIterator() and getValues() instead of only exposing binary APIs and pushing the responsibilities of serialization and deserialization to the client. Similarly, the DiskStore put() methods accepted a `StorageLevel` parameter even though the disk store can only store blocks in one form.

As part of a larger BlockManager interface cleanup, this patch remove the BlockStore interface and refines the MemoryStore and DiskStore interfaces to reflect more narrow sets of responsibilities for those components. Some of the benefits of this interface cleanup are reflected in simplifications to several unit tests to eliminate now-unnecessary mocking, significant simplification of the BlockManager's `getLocal()` and `doPut()` methods, and a narrower API between the MemoryStore and DiskStore.

JoshRosen committed:
Delete the BlockStore interface.

JoshRosen committed:
Remove unused DiskStore.getBytes() overloads.

JoshRosen committed:
DiskStore.getBytes() never returns None, so it shouldn't return an Option.

JoshRosen committed:
Simplify DiskStore.putIterator's return type.

JoshRosen committed:
Remove MemoryStore.putIterator() overload.

JoshRosen committed:
DiskStore put() methods don't need to take a StorageLevel.

JoshRosen committed:
Factor common error-handling code in DiskStore.put*() into helper function.

JoshRosen committed:
Remove DiskStore.putIterator().

JoshRosen committed:
Remove DiskStore's dependency on BlockManager.

JoshRosen committed:
Minor simplifications in DiskStoreSuite test.

JoshRosen committed:
Remove outdated comment in DiskBlockManager.

JoshRosen committed:
Remove DiskStore's dependency on BlockManager.

JoshRosen committed:
Shorten period of holding memoryManager lock.

JoshRosen committed:
MemoryStore.put() no longer handles dropping to disk.

This is now handled by the caller.

JoshRosen committed:
MemoryStore.putBytes() shouldn't perform deserialization.

JoshRosen committed:
MemoryStore should take its own conf, not obtain it from BlockManager.

JoshRosen committed:
Move MemoryManager into new o.a.s.storage.memory package

JoshRosen committed:
getBytes() and getValues() no longer implicitly serialize / deserialize.

JoshRosen committed:
Split doGetLocal() and getLocal() into smaller, simpler methods.

SparkQA commented:
**[Test build #52506 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52506/consoleFull)** for PR 11534 at commit [`7d2aba3`](https://github.com/apache/spark/commit/7d2aba3997ba3770c45585daee4e023a172a4f50).

SparkQA commented:
**[Test build #52506 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52506/consoleFull)** for PR 11534 at commit [`7d2aba3`](https://github.com/apache/spark/commit/7d2aba3997ba3770c45585daee4e023a172a4f50).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52506/
Test FAILed.

JoshRosen committed:
Fix scalastyle violations.

SparkQA commented:
**[Test build #52521 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52521/consoleFull)** for PR 11534 at commit [`18c07e6`](https://github.com/apache/spark/commit/18c07e6da858be167500220732d3bb4565e98df2).

JoshRosen commented:
There seems to be a lot of 

```
12:54:15.739 WARN org.apache.spark.storage.BlockManager: Block rdd_38548_0 already exists on this machine; not re-adding it
12:54:15.742 ERROR org.apache.spark.executor.Executor: 1 block locks were not released by TID = 103091:
[rdd_38548_0]
```

entries in the test log during HiveCompatibilitySuite. I'll try to investigate to see if I can spot what's going on. This isn't leading to test failures but likely indicates a problem somewhere.

JoshRosen committed:
Fix leaked lock in getOrElseUpdate() when block already exists.

JoshRosen commented:
Spotted the problem: we need an extra `releaseLock()` call in `getOrElseUpdate`'s "block already exists" branch; see 9dc510c. I can spin this change off into a smaller patch against master in case we want to clean up the log noise sooner.

SparkQA commented:
**[Test build #52522 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52522/consoleFull)** for PR 11534 at commit [`9dc510c`](https://github.com/apache/spark/commit/9dc510ca373ff02d71bf54b952396b0c1284ef0f).

SparkQA commented:
**[Test build #52521 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52521/consoleFull)** for PR 11534 at commit [`18c07e6`](https://github.com/apache/spark/commit/18c07e6da858be167500220732d3bb4565e98df2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52521/
Test PASSed.

SparkQA commented:
**[Test build #52522 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52522/consoleFull)** for PR 11534 at commit [`9dc510c`](https://github.com/apache/spark/commit/9dc510ca373ff02d71bf54b952396b0c1284ef0f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52522/
Test PASSed.

nongli commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -401,11 +403,49 @@ private[spark] class BlockManager(
+          val iterToReturn: Iterator[Any] = {
+            if (level.deserialized) {
+              // Cache the values before returning them
+              memoryStore.putIterator(blockId, valuesFromDisk, level) match {
```
This seems like confusing putIterator semantics...

JoshRosen replied:
How would you simplify it? We have to return an iterator on failure because of the one-shot iterator semantics.

Do you think that we should just throw away the partially-unrolled iterator returned from `putIterator()` and perform another disk scan instead?

nongli commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -422,108 +462,52 @@ private[spark] class BlockManager(
-        logDebug(s"Level for block $blockId is $level")
-
-        // Look for the block in memory
+  private def doGetLocalBytes(blockId: BlockId, info: BlockInfo): ByteBuffer = {
```
Comment that this is called with the read lock taken.

JoshRosen replied:
Done.

nongli commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -422,108 +462,52 @@ private[spark] class BlockManager(
+  private def doGetLocalBytes(blockId: BlockId, info: BlockInfo): ByteBuffer = {
+    val level = info.level
+    logDebug(s"Level for block $blockId is $level")
+    if (level.deserialized) {
```
Is this backwards? If the level is deserialized we serialize it? comment.

JoshRosen replied:
`level-deserialized` specifies that this block's in-memory values will be stored as Java objects. For a block at this storage level, there are three possibilities:

1. it's stored as java objects in memory
2. it's stored as bytes on disk
3. an error has occurred and the block is missing

The logic inside of this `if` case tries to check case 2 ahead of case 1 so that we avoid unnecessary serialization if a serialized copy already exists on disk.

Were you confused about the fact that we can call `getLocalBytes()` on a block which is cached as java objects?

nongli commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -744,7 +732,7 @@ private[spark] class BlockManager(
       tellMaster: Boolean = true,
       effectiveStorageLevel: Option[StorageLevel] = None,
-      keepReadLock: Boolean = false): Option[PutResult] = {
+      keepReadLock: Boolean = false): Option[Option[Iterator[Any]]] = {
```
Why do levels of options? What does it mean if the outer is set but the inner is not

JoshRosen replied:
This was discussed extensively on https://github.com/apache/spark/pull/11502 and will be fixed once I merge that PR.

nongli commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -801,43 +780,63 @@ private[spark] class BlockManager(
-    var result: PutResult = null
+    var iteratorFromFailedMemoryStorePut: Option[Iterator[Any]] = None
 
     putBlockInfo.synchronized {
```
What is this synchronized protecting? You have an exclusive write lock now no?

JoshRosen replied:
I think this was a carryover from the very old code where we synchronized on the BlockInfo objects themselves instead of using our new lock interface. I'll remove this to clean things up.

nongli commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -17,125 +17,100 @@
-  }
-
-  override def remove(blockId: BlockId): Boolean = {
+  def remove(blockId: BlockId): Boolean = {
```
is this return code useful?

JoshRosen replied:
Removals from the stores are idempotent, so the return value lets the caller check whether the removal actually deleted a block (see usage in `BlockManager.removeBlock()`).

nongli commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -17,125 +17,100 @@
 import org.apache.spark.util.Utils
 
 /**
  * Stores BlockManager blocks on disk.
```
This not thread safe correct?

JoshRosen replied:
Yes, that's right. Thread-safety is ensured by locking at higher levels of the system.

nongli commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -34,13 +35,15 @@ private case class MemoryEntry(value: Any, size: Long, deserialized: Boolean)
   // acquiring or releasing unroll memory, must be synchronized on `memoryManager`!
 
-  private val conf = blockManager.conf
   private val entries = new LinkedHashMap[BlockId, MemoryEntry](32, 0.75f, true)
```
Comment on thread safety. Why do you need locks for entries?

JoshRosen replied:
Good catch. In the old code, there were multiple places that called into the MemoryStore, some in the BlockManager and others in the CahceManager, so I think we needed these locks to ensure thread-safety. Now, though, I think that all accesses should be coming from the BlockManager so I think we can remove this locking and rely on the caller holding the appropriate BlockManager locks instead.

JoshRosen replied:
Actually, nevermind: we _do_ need this synchronization to guarantee the memory safety / consistency of the bookkeeping structures themselves. This synchronization is distinct from the per-block locking that's done higher up the stack.

nongli replied:
Hmm okay. DiskStore also needs to be thread safe right? It just happens it doesn't maintain any state right now that's interesting. I think it's worth commenting so people modifying this know what the expected properties are.

nongli commented:
Overall changes look good. Need to take another pass or two to understand this code better.

JoshRosen commented:
FYI I rebased this to fix the merge conflicts, so the diff + commit log is now smaller.

SparkQA commented:
**[Test build #52638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52638/consoleFull)** for PR 11534 at commit [`04f8d15`](https://github.com/apache/spark/commit/04f8d15b8877a508fba70ba1ee118bf9d199973b).

JoshRosen commented:
Splitting up `doPut()` into two smaller methods, one concerned with putting bytes and another for putting iterators of objects, would let me simplify its return type and might lead to other simplifications (e.g. by eliminating impossible error-handling cases).  I might try this tomorrow.

SparkQA commented:
**[Test build #52638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52638/consoleFull)** for PR 11534 at commit [`04f8d15`](https://github.com/apache/spark/commit/04f8d15b8877a508fba70ba1ee118bf9d199973b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52638/
Test PASSed.

SparkQA commented:
**[Test build #52679 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52679/consoleFull)** for PR 11534 at commit [`988f003`](https://github.com/apache/spark/commit/988f00393676eabfc11e665f20f9ce26388e4c11).

JoshRosen committed:
Document lock requirements of doGetLocalBytes

JoshRosen committed:
Add clarifying comment to doGetLocalBytes()

JoshRosen committed:
Remove unnecessary putBlockInfo.synchronized call

SparkQA commented:
**[Test build #52683 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52683/consoleFull)** for PR 11534 at commit [`14857b3`](https://github.com/apache/spark/commit/14857b3e53d1b496b7ff07099c53ab2d775f950a).

SparkQA commented:
**[Test build #52679 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52679/consoleFull)** for PR 11534 at commit [`988f003`](https://github.com/apache/spark/commit/988f00393676eabfc11e665f20f9ce26388e4c11).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52679/
Test PASSed.

SparkQA commented:
**[Test build #52683 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52683/consoleFull)** for PR 11534 at commit [`14857b3`](https://github.com/apache/spark/commit/14857b3e53d1b496b7ff07099c53ab2d775f950a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52683/
Test FAILed.

JoshRosen committed:
Remove effectiveStorageLevel from put() APIs.

SparkQA commented:
**[Test build #52702 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52702/consoleFull)** for PR 11534 at commit [`92c5125`](https://github.com/apache/spark/commit/92c5125f2f4736335971e779fc39e9fa74f8c310).

JoshRosen committed:
Split doPut() into doPutBytes() and doPutIterator().

JoshRosen committed:
Remove unreachable level == StorageLevel.NONE case.

This is unreachable because we check whether level.isValid earlier in the same method.

SparkQA commented:
**[Test build #52704 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52704/consoleFull)** for PR 11534 at commit [`7a08a17`](https://github.com/apache/spark/commit/7a08a179f8951abbdcb7e70f6bfb53821fbc7352).

JoshRosen committed:
Fix statement without side-effects.

SparkQA commented:
**[Test build #52705 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52705/consoleFull)** for PR 11534 at commit [`a16276e`](https://github.com/apache/spark/commit/a16276e1cfb92fba88f2963b0d918aa14f2500a4).

JoshRosen committed:
Minor comment reword.

JoshRosen commented:
I went ahead and split up `doPut()`, which greatly simplified the return types.

SparkQA commented:
**[Test build #52707 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52707/consoleFull)** for PR 11534 at commit [`66796b5`](https://github.com/apache/spark/commit/66796b5bf89ddedc9644b9f5692441293c0c0aaa).

SparkQA commented:
**[Test build #52702 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52702/consoleFull)** for PR 11534 at commit [`92c5125`](https://github.com/apache/spark/commit/92c5125f2f4736335971e779fc39e9fa74f8c310).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52702/
Test PASSed.

SparkQA commented:
**[Test build #52705 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52705/consoleFull)** for PR 11534 at commit [`a16276e`](https://github.com/apache/spark/commit/a16276e1cfb92fba88f2963b0d918aa14f2500a4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52705/
Test FAILed.

SparkQA commented:
**[Test build #52704 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52704/consoleFull)** for PR 11534 at commit [`7a08a17`](https://github.com/apache/spark/commit/7a08a179f8951abbdcb7e70f6bfb53821fbc7352).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52704/
Test PASSed.

SparkQA commented:
**[Test build #52707 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52707/consoleFull)** for PR 11534 at commit [`66796b5`](https://github.com/apache/spark/commit/66796b5bf89ddedc9644b9f5692441293c0c0aaa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52707/
Test PASSed.

nongli commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -17,112 +17,100 @@
-    val file = diskManager.getFile(blockId)
-    val outputStream = new FileOutputStream(file)
-    try {
+  def putBytes(blockId: BlockId, _bytes: ByteBuffer): Unit = {
```
Comment function and guarantees on _bytes on return.

nongli commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -789,117 +752,193 @@ private[spark] class BlockManager(
+            if (level.useDisk) {
+              logWarning(s"Persisting block $blockId to disk instead.")
+              diskStore.put(blockId) { fileOutputStream =>
+                dataSerializeStream(blockId, fileOutputStream, iter)
```
Does this mean this block is now split between the memStore and the diskStore?

JoshRosen replied:
Nope, the entire block is now on disk because `iter` contains all of the elements of the original input `iterator()`.

nongli commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -789,117 +752,193 @@ private[spark] class BlockManager(
+   * @param keepReadLock if true, this method will hold the read lock when it returns (even if the
+   *                     block already exists). If false, this method will hold no locks when it
+   *                     returns.
+   * @return None if the block was already present or if the put succeeded, or Some(iterator)
```
Can you comment on any properties of the returned iterator in this case?

Just so I understand, if this returns None, we've walked the input iterator to the end. If it returns an iterator, it has been advanced arbitrarily far along?

JoshRosen replied:
Yep, that's right. The returned iterator, if present, contains the same elements as the original iterator.

This is going to become a little clearer after yet another followup patch to refactor the equivalent memorystore code.

nongli commented:
LGTM

JoshRosen committed:
Merge doGetRemote() with getRemoteBytes()

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -407,11 +399,49 @@ private[spark] class BlockManager(
+          val ci = CompletionIterator[Any, Iterator[Any]](iterToReturn, releaseLock(blockId))
+          Some(new BlockResult(ci, DataReadMethod.Disk, info.size))
+        } else {
+          releaseLock(blockId)
```
do we need to release in finally?

JoshRosen replied:
Nope; we want the caller to hold a read lock if this method returns successfully.

andrewor14 replied:
can you add a comment in the javadoc on when the lock will be released?

SparkQA commented:
**[Test build #52777 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52777/consoleFull)** for PR 11534 at commit [`94d2bd6`](https://github.com/apache/spark/commit/94d2bd65f4fc70e75ddff4fe70c8fc4c68b481af).

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -428,77 +458,46 @@ private[spark] class BlockManager(
+          None
+        case Some(info) =>
+          Some(doGetLocalBytes(blockId, info))
+      }
```
don't know if we care about the debug message, but a nicer way to write this is:
```
blockInfoManager.lockForReading(blockId).map { info => doGetLocalBytes(blockId, info) }
```

JoshRosen replied:
I considered this but thought that doing so would be more confusing to readers.

andrewor14 replied:
I think it's pretty clear, but we don't have to do it

JoshRosen replied:
Done.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -407,11 +399,49 @@ private[spark] class BlockManager(
+        } else if (level.useDisk && diskStore.contains(blockId)) {
+          val valuesFromDisk = dataDeserialize(blockId, diskStore.getBytes(blockId))
+          val iterToReturn: Iterator[Any] = {
+            if (level.deserialized) {
```
There's not really any reason why we should put it back in memory only for the deserialized case. If I have `MEM_AND_DISK_SER` then this won't even try to the bytes in memory.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -428,77 +458,46 @@ private[spark] class BlockManager(
+    if (level.deserialized) {
+      // Try to avoid expensive serialization by reading a pre-serialized copy from disk:
+      if (level.useDisk && diskStore.contains(blockId)) {
+        diskStore.getBytes(blockId)
```
I would add a comment here saying don't bother putting it back in memory here because we'd have to deserialize the data and that might be slow.

JoshRosen replied:
To be more specific, we don't put the bytes into memory here because a) the block is stored at a deserialized level, so per our contract it should _only_ be stored as values in memory and not bytes, and b) the caller of this method (`doGetLocalBytes()`) has requested bytes and not values, so deserializing them is just extra work that's likely not to have a payoff.

andrewor14 replied:
yeah, that's what I meant

JoshRosen replied:
Comment added.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -548,16 +521,9 @@ private[spark] class BlockManager(
    * This does not acquire a lock on this block in this JVM.
    */
   def getRemote(blockId: BlockId): Option[BlockResult] = {
```
can you rename this `getRemoteValues` to be consistent? Now we have
```
getLocalValues
getLocalBytes
getRemote
getRemoteBytes
```
so this one stands out

JoshRosen replied:
Done.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -428,77 +458,46 @@ private[spark] class BlockManager(
-        None
-      case Some(info) =>
-        doGetLocal(blockId, info, asBlockResult)
+      blockInfoManager.lockForReading(blockId) match {
```
need to clarify in javadocs the locking semantics of `getLocalValues` and `getLocalBytes`. In particular, both of these methods acquire the read lock. In the iterator case we release the lock when the iterator is drained. In the bytes case the caller is responsible for releasing the locks. In both cases, if we couldn't get the blocks for some reason, we release the lock and throw an exception instead. (Is that right?)

JoshRosen replied:
That's correct.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -789,117 +742,193 @@ private[spark] class BlockManager(
+    logDebug("Put block %s locally took %s".format(blockId, Utils.getUsedTimeMs(startTimeMs)))
+    if (level.replication > 1 && blockWasSuccessfullyStored) {
+      val remoteStartTime = System.currentTimeMillis
+      val bytesToReplicate = doGetLocalBytes(blockId, putBlockInfo)
```
if `keepReadLock` is false we don't actually hold the read lock here! That's not what `doGetLocalBytes` assumes.

JoshRosen replied:
I'll just move this code inside of the `try` block.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -701,16 +666,15 @@ private[spark] class BlockManager(
-    result == DoPutSucceeded
+      tellMaster)
+    // If doPut() didn't hand work back to us, then block already existed or was successfully stored
+    maybeIterator.isEmpty
```
I would call this `failedIterator` or something to be more specific.

JoshRosen replied:
I just removed the variable, since the entire line and the `isEmpty` fit together.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -701,16 +666,15 @@ private[spark] class BlockManager(
-      tellMaster,
-      effectiveStorageLevel)
-    result == DoPutSucceeded
+      tellMaster)
```
nit: just put these in 1 line. It fits.

JoshRosen replied:
Done.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -916,11 +945,9 @@ private[spark] class BlockManager(
+      None
     } else {
-      DoPutBytesFailed
+      Some(iteratorFromFailedMemoryStorePut.get)
```
I would just return `iteratorFromFailedMemoryStorePut` here directly. Doing the `Some(opt.get)` is a strange pattern. If you want you can also add an `assert(blockWasSuccessfullyStored == iteratorFromFailedMemoryStorePut.isEmpty)`.

JoshRosen replied:
Done.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -789,117 +742,193 @@ private[spark] class BlockManager(
+        if (!keepReadLock) {
+          // lockNewBlockForWriting returned a read lock on the existing block, so we must free it:
+          releaseLock(blockId)
+        }
```
This is the only hesitation about this patch that I have. There's now a non-trivial amount of code that's duplicated in `doPutIterator` and `doPutBytes`. Is there any way to abstract it? Otherwise I'm not sure if we have a net reduction in complexity here. :/

JoshRosen replied:
I've tried factoring it out into a common method but that method has a weird name and signature and the size of the comment needed to describe it meant that we didn't end up saving much space. Basically, I'd like to defer that cleanup to later since it seems non-essential and I think this is already more understandable than the old code.

JoshRosen committed:
getRemote -> getRemoteValues()

JoshRosen committed:
Move putIterator() block replication code into try block so that we hold read lock.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -789,117 +742,193 @@ private[spark] class BlockManager(
+        // We will drop it to disk later if the memory store can't hold it.
+        val putSucceeded = if (level.deserialized) {
+          val values = dataDeserialize(blockId, bytes.duplicate())
+          memoryStore.putIterator(blockId, values, level).isRight
```
This is in `doPutBytes`. Similar to another comment I had above, there's no reason why we shouldn't attempt to drop it to disk here if it's left. Right now if `memoryStore.putIterator` doesn't fit in memory then we sometimes drop it to disk but sometimes not. That's confusing.

JoshRosen replied:
If `isRight` here returns `false` then in the `if (!putSucceeded && level.useDisk) {` branch four lines down we'll try to put the original bytes to disk.

andrewor14 replied:
ah ok, never mind

SparkQA commented:
**[Test build #52787 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52787/consoleFull)** for PR 11534 at commit [`e3a4c22`](https://github.com/apache/spark/commit/e3a4c223fcfd943266590a479f7e43d78bafabd5).

JoshRosen committed:
Simplify return at end of doPutIterator()

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala:
```diff
@@ -26,18 +26,14 @@ import org.apache.spark.util.{ShutdownHookManager, Utils}
  */
-private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkConf)
-  extends Logging {
+private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolean) extends Logging {
```
nice!

SparkQA commented:
**[Test build #52791 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52791/consoleFull)** for PR 11534 at commit [`0147471`](https://github.com/apache/spark/commit/01474717ac30d472e24ebdb35e0ea5f8b9570e3c).

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -318,19 +291,26 @@ private[spark] class MemoryStore(blockManager: BlockManager, memoryManager: Memo
     blockId.asRDDId.map(_.rddId)
   }
 
+  private def acquireStorageMemory(blockId: BlockId, size: Long): Boolean = {
```
IIUC this is only called in 1 place. Why bother abstracting it? It's just kind of weird to see that it's releasing pending unroll memory outside the context of putting a block. I actually think this fits better where it used to be.

JoshRosen replied:
Sure, I can move it back. This method is going away anyways in my next patch, though.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -114,89 +106,70 @@ private[spark] class MemoryStore(blockManager: BlockManager, memoryManager: Memo
-   * There may not be enough space to fully unroll the iterator in memory, in which case we
-   * optionally drop the values to disk if
-   *   (1) the block's storage level specifies useDisk, and
-   *   (2) `allowPersistToDisk` is true.
```
great! I hated this flag.

JoshRosen committed:
Move acquireStorageMemory back to old call site.

JoshRosen committed:
Fix minor formatting issue in putIterator().

SparkQA commented:
**[Test build #52792 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52792/consoleFull)** for PR 11534 at commit [`84e4d76`](https://github.com/apache/spark/commit/84e4d768dfc52a7a610cb935c838e76eb28fbebf).

andrewor14 commented:
@JoshRosen This looks great. The only concern I have is the duplicate code in `doPutIterator` and `doPutBytes`. In the future if we find a bug in one of them and forget to fix it in the other place things will get really hard to maintain. I do agree on the general idea of separating the two instead of having one giant monolithic `doPut` method, but I really think we need to fix the duplicate code here and I would prefer to do it in this patch.

SparkQA commented:
**[Test build #52794 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52794/consoleFull)** for PR 11534 at commit [`e2ee2e2`](https://github.com/apache/spark/commit/e2ee2e2af503f759a1b152d1dc9e9c7be3616d94).

JoshRosen committed:
Add comment to explain why we don't cache deserialized blocks' bytes in memory.

JoshRosen commented:
@andrewor14, let me take another shot at reducing that duplication and I'll get back to you.

JoshRosen committed:
Simplify code in getLocalBytes

SparkQA commented:
**[Test build #52797 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52797/consoleFull)** for PR 11534 at commit [`08bc93b`](https://github.com/apache/spark/commit/08bc93b4f0734c7739534a47388238beab39efe8).

JoshRosen committed:
Tautology

SparkQA commented:
**[Test build #52777 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52777/consoleFull)** for PR 11534 at commit [`94d2bd6`](https://github.com/apache/spark/commit/94d2bd65f4fc70e75ddff4fe70c8fc4c68b481af).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52777/
Test PASSed.

JoshRosen committed:
Move async replication wait code into try block

SparkQA commented:
**[Test build #52792 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52792/consoleFull)** for PR 11534 at commit [`84e4d76`](https://github.com/apache/spark/commit/84e4d768dfc52a7a610cb935c838e76eb28fbebf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52792/
Test PASSed.

SparkQA commented:
**[Test build #52794 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52794/consoleFull)** for PR 11534 at commit [`e2ee2e2`](https://github.com/apache/spark/commit/e2ee2e2af503f759a1b152d1dc9e9c7be3616d94).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52794/
Test PASSed.

SparkQA commented:
**[Test build #52791 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52791/consoleFull)** for PR 11534 at commit [`0147471`](https://github.com/apache/spark/commit/01474717ac30d472e24ebdb35e0ea5f8b9570e3c).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52791/
Test FAILed.

SparkQA commented:
**[Test build #52797 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52797/consoleFull)** for PR 11534 at commit [`08bc93b`](https://github.com/apache/spark/commit/08bc93b4f0734c7739534a47388238beab39efe8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52797/
Test PASSed.

SparkQA commented:
**[Test build #52787 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52787/consoleFull)** for PR 11534 at commit [`e3a4c22`](https://github.com/apache/spark/commit/e3a4c223fcfd943266590a479f7e43d78bafabd5).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52787/
Test FAILed.

JoshRosen committed:
Factor out duplicated code from the doPut() methods.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -739,46 +696,105 @@ private[spark] class BlockManager(
+      if (blockWasSuccessfullyStored) {
+        None
+      } else {
+        Some(bytes)
```
I don't think we ever use this anywhere. It's simpler to just return `blockWasSuccessfullyStored` here so this method can just return the result of `doPut` directly instead of `doPut(...) { ... }.isEmpty`

andrewor14 commented:
LGTM, can't wait to merge this!

JoshRosen committed:
Always try to cache bytes in memory when reading serialized blocks from disk.

JoshRosen committed:
Style cleanup.

SparkQA commented:
**[Test build #52850 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52850/consoleFull)** for PR 11534 at commit [`2752655`](https://github.com/apache/spark/commit/2752655518c9907dd2c9195f7da8cff28824d71e).

SparkQA commented:
**[Test build #52845 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52845/consoleFull)** for PR 11534 at commit [`2e8b531`](https://github.com/apache/spark/commit/2e8b531ff6393b6d022e36b53f11befd3576f0b5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52845/
Test FAILed.

SparkQA commented:
**[Test build #52850 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52850/consoleFull)** for PR 11534 at commit [`2752655`](https://github.com/apache/spark/commit/2752655518c9907dd2c9195f7da8cff28824d71e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52850/
Test PASSed.

andrewor14 commented:
Merged into master.
